source/log_channel.c

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/io/log_channel.h>
+
+#include <aws/common/condition_variable.h>
+#include <aws/common/mutex.h>
+#include <aws/common/string.h>
+#include <aws/common/thread.h>
+#include <aws/io/log_writer.h>
+
+/*
+ * Basic channel implementations - synchronized foreground, synchronized background
+ */
+
+struct aws_log_foreground_channel {
+    struct aws_mutex sync;
+};
+
+static int s_foreground_channel_send_fn(
+    struct aws_log_channel *channel,
+    struct aws_string *log_line) {
+
+    struct aws_log_foreground_channel *impl = (struct aws_log_foreground_channel *) channel->impl;
+
+    aws_mutex_lock(&impl->sync);
+    (*channel->writer->vtable->write_fn)(channel->writer, log_line);
+    aws_mutex_unlock(&impl->sync);
+
+    /*
+     * send is considered a transfer of ownership.  write is not a transfer of ownership.
+     * So it's always the channel's responsibility to clean up all log lines that enter
+     * it as soon as they are no longer needed.
+     */
+    aws_string_destroy(log_line);
```

Comment:
I guess as a default this is fine, but we'll probably want to re-use this bad boy, or have 1 per thread if we use this impl, to avoid actually freeing them

Reply:
Re: this comment as well as the next

My current thought is to use a custom allocator based on a ringlike memory buffer (per calling thread) for the formatted strings.  By making it per thread, the allocate and frees from each pool stay exactly in order so we could get away with a couple of pointer offsets as all the state we need to allocate/free, with allocation and free being lockless against a couple of atomic values, avoiding the need for a much more complex free list/malloc implementation.  As long as the background thread kept up with the strings being thrown at it, allocation and freeing would be almost completely trivial (tls access + a few atomic integer operations).

We'd need to handle the case when the ring buffer didn't have enough room or the request was too big for it. The easiest thing to do would be just go to the heap.  If we make the ring buffer size tunable, then configuration would let you optimize things to where you "almost always" end up using the ring buffer in practice.

So allocating a string would fall back to the heap when the ring buffer can't handle the request, and freeing would check if the pointer is in the ring buffer and use the heap (well the formatter's allocator) if it's not.

Reply:
That sounds good, since the string contains an allocator*, it could be re-routed. Sounds good to me.